### ISSUE TITLE
Remove object-based permissions

### ISSUE THREAD

**Issue by annevk** [w3c/permissions#42]
In all implementations there are only string-based permissions. I don't see why we should make the API higher-level and more complicated than what browsers actually do.

**Comment by marcoscaceres** [comment]
@annevk, you are suggesting for basic types we just use [PermissionName](https://w3c.github.io/permissions/#enumdef-permissionname), right? Name is always required, so that would make sense.  

And I guess in relation to https://github.com/w3c/permissions/issues/59, the there could be a second argument for the "PermissionDescriptor" part (as what  @jan-ivar seems to suggest). That would avoid the need for any dictionaries to extend PermissionDescriptor.

so, something like:

```JS
navigator.permissions.query("some-permission", {some: "details"})
```

**Comment by marcoscaceres** [comment]
A more concrete example:
```JS
navigator.permissions.query('camera', { deviceId }).then(...)
```

**Comment by jyasskin** [comment]
@mounirlamouri Any thoughts on making this a 2-parameter function? Since it's shipped, we'd need to allow `{name: 'camera'}` for the first argument, but we could also allow the more convenient form.

**Comment by marcoscaceres** [comment]
@jyasskin, if the claim in https://github.com/w3c/permissions/issues/59 is correct, then allowing `name`, in theory, wont work in an interoperable manner. Is that correct, @jan-ivar?

**Comment by jyasskin** [comment]
#59 just means that we should specify the interface slightly differently, or that implementations will have to use different WebIDL than we have in the spec. It's not a problem with the overall shape of the interface.

**Comment by marcoscaceres** [comment]
> #59 just means that we should specify the interface slightly differently, or that implementations will have to use different WebIDL than we have in the spec.

@jyasskin or @jan-ivar, sorry, I don't understand why tho. Is it possible to provide a brief explanation or point me to the right place in the WebIDL spec so I can understand what doesn't work interoperably? 

Also, should we do one argument, that allows `PermissionName` or `PermissionDescriptor`. Having a single argument could work similarly to how `fetch()` allows either a `Request` or `USVString` as `RequestInfo`.  

As in: 
```JS
navigator.permissions.query("geolocation");
navigator.permissions.query({name: "geolocation", otherStuff: ""});
```

**Comment by alvestrand** [comment]
I'm peculiar, but I find permissions.query({name: "geolocation"}) easy to read.

**Comment by marcoscaceres** [comment]
> I'm peculiar, but I find permissions.query({name: "geolocation"}) easy to read.

It's not really an ergonomic issues, but if we support both enum and dictionary as input, we get the ergonomics (+backwards compat) for free.

**Comment by jan-ivar** [comment]
@marcoscaceres sorry for the delay. See https://github.com/w3c/webrtc-pc/issues/275#issuecomment-134979978 for relevant WebIDL links and the discussion that follows, for why dictionaries are not polymorphic.

@jyasskin As written, [the WebIDL in the spec](https://w3c.github.io/permissions/#permissions-interface) is incorrect for all three methods. I think we need valid WebIDL here if we're going to have WebIDL at all, otherwise it is misleading.

Specifically: An implementation using a WebIDL compiler would not be able to - as the spec's normative prose dictates - "convert the underlying ECMAScript object to the permission descriptor type dictionary as described in [WEBIDL]", because the underlying ECMAScript object by definition is not available (*permissionDesc* is a secure copy of the base dictionary only), unless we were to do:

```js
[Exposed=(Window,Worker)]
interface Permissions {
  Promise<object> query(object permissionDesc);

  Promise<object> request(object permissionDesc);

  Promise<object> revoke(object permissionDesc);
};
```

or something like it (`object` has terrible security characteristics). The spec needs to be clear here.

**Comment by jyasskin** [comment]
@jan-ivar Can you link to an explanation of the security claim? Otherwise, using `object` instead of `PermissionDescriptor`, along with an explicit type check in prose, is exactly what "specify the interface slightly differently" meant.

**Comment by jyasskin** [comment]
For @marcoscaceres, WebIDL doesn't define the "underlying ECMAScript object" operation that our functions rely on. Dictionaries are [passed as copies](http://heycam.github.io/webidl/#es-dictionary), as @jan-ivar said, so we'd need to store an extra link in WebIDL to even make it possible. But it's not a big change to use `object` instead.

**Comment by annevk** [comment]
I feel like this isn't really getting to the point made in OP.

**Comment by jyasskin** [comment]
@annevk Indeed, we've gotten side-tracked by #59, which is easily fixed. You can see with the Bluetooth permission that there are uses for object-based permissions: are you happy with the current API shape, or would you rather a (DOMString or object) argument type, or something else?

**Comment by annevk** [comment]
I have not seen the Bluetooth stuff. There's also still lack of agreement around that API aiui.

**Comment by jyasskin** [comment]
@annevk Please do look at how [Web Bluetooth integrates with the Permissions API](https://webbluetoothcg.github.io/web-bluetooth/#dom-permissionname-bluetooth). Even if that API turns out to be too much of a security risk to ship, the shape of the permissions API should allow other similar features. The TAG actually asked for the object parameters in [its review](https://github.com/w3ctag/spec-reviews/blob/master/2015/03/Permissions.md#end-notes).

**Comment by annevk** [comment]
Is there going to be different UX for all those parameters? TAG examples are all boolean in nature. Also, complexity can always be added later. Starting simple does not prevent that.

**Comment by jyasskin** [comment]
Taking [`enableHighAccuracy`](https://github.com/w3ctag/spec-reviews/blob/master/2015/03/Permissions.md#insufficient-reflection) as an example, there might be a difference in the prompt (UX) to express whether the page wants coarse or expensive results, and developers have to be aware of the `enableHighAccuracy` field in geolocation's `PermissionDescriptor` in order to use it.

I think Bluetooth could get away with just `query("bluetooth")` by returning the full list of allowed `BluetoothDevice` objects, and making the developer filter them. That would be more difficult with Geolocation, MIDI, and Push, where the different levels of the API could be most combinations of `"granted"`, `"prompt"`, or `"denied"`. Bluetooth couldn't get away with just `request("bluetooth")`, because of the need to filter the prompt.

**Comment by annevk** [comment]
Yeah, I meant that for booleans like high accuracy you could just have "geolocation-precise". It's not terribly clear to me user agents would actually start storing keys as objects here. The API seems much higher-level than what we'd implement under the hood.

**Comment by jyasskin** [comment]
Ah, right, and for midi we'd have <code>"midi-[sysex](https://webaudio.github.io/web-midi-api/#widl-MIDIOptions-sysex)-[software](https://webaudio.github.io/web-midi-api/#widl-MIDIOptions-software)"</code>, etc. That's plausible, but I think it gets klunky when you have more than one option.

UAs do have to do more than store a string here: we [should forbid](https://w3c.github.io/permissions/#permissiondescriptor-stronger-than) weird behaviors like `"geolocation"` being denied while `"geolocation-precise"` is granted. However, that doesn't imply anything for the shape of the API.

@alvestrand has a different vision for how `query()` and `request()` should work than I do (his `query()` stays orthogonal to [`enumerateDevices()`](https://w3c.github.io/mediacapture-main/#enumerating-devices) instead of obsoleting it), but even his more tightly scoped view involves a string in the descriptor dictionary: https://w3c.github.io/permissions/#media-devices.

**Comment by jan-ivar** [comment]
@jyasskin How would `query()` obsolete `enumerateDevices()`?

**Comment by alvestrand** [comment]
The important reason for using an object selector rather than a string selector is that the API needs to support querying / requesting permission for multiple devices.
These can, in theory, be named by strings ("camera-xyzaerg"), since the camera ID is a string, but I dislike going to string parsing when we have an object model available.
